Refactor code that deploys into testnet

itamararjuan opened the pull request:
This PR adds code that deals with deploying into the CI testnet in a manner that follows our approach in recent months of having a centralized S3 bucket that holds information about which virtual chains Boyar should deploy.

The code downloads and modifies the CI's testnet S3 bucket central config and updates the version to deploy for all the chains available there and re-uploads that file into S3.

In addition, the code aims to check that the version was indeed deployed on all provided nodes and that the block height is progressing in all chains. then and only then will the step of deployment to testnet completes successfully and e2e tests can run safely against said chains.

The code also fixes a problem with parsing the nightly junit xml stats as well.

electricmonk commented:
@netoneko please review

netoneko commented on .circleci/check-testnet-deployment.js:
```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env node
```
We don't need this code. It can be solved by importing nebula and using metrics-related methods from there.

itamararjuan replied:
After going through your metrics related code in Nebula  - I am not quite sure it's the same.
The method `waitForSync()` waits for 1 endpoint to sync. My code here is not testing for this.
Syncing can happen also with mismatched versions and that's not what I'm aiming for.
I'm not moving forward until I see the network is completely updated to the new version as a deployment should be.

I agree that we can replace my code for `eventuallyClosingBlocks()` and use your `waitForSync()`
What do you think ?

netoneko commented on .circleci/deploy-testnet.sh:
```diff
@@ -1,16 +1,27 @@
+node .circleci/testnet-deploy-tag.js $COMMIT_HASH
 
-export COMMIT_HASH=$(./docker/hash.sh)
+aws s3 cp --acl public-read config.json s3://boyar-ci/boyar/config.json
```
Please use nebula instead. We have all the API that is needed for that.

itamararjuan replied:
The code in Nebula can generate a complete configuration and requires the complete set of network  JSON files. 
If I do that here - I need to copy or add a git submodule of testnet configuration in here (which is a private repo)

I was aiming to try and only change the version deployed to the chains in the file instead of changing the entire file. I can do that - let me know how do you want to solve sharing the JSONs required to run these APIs in nebula.

The `aws s3 cp` calls are still required as far as I can remember.

netoneko replied:
Ok, let's keep it that way for now.

netoneko commented on .circleci/testnet-deploy-tag.js:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env node
```
Same about nebula.

itamararjuan replied:
This will be decided once you answer my comment regarding how to share the testnet configuration.
Perhaps we shall move the configuration of the CI testnet over to this repo and save us from the complexity of sharing a piece of a private repo.

netoneko commented on test.flakiness.sh:
```diff
@@ -18,6 +18,10 @@ if [[ "${LAST_COMMIT_MESSAGE}" == *"#extraflaky"* ]]; then
 fi
 
 if [[ $1 == "NIGHTLY" ]]; then
+    curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.11/install.sh | bash
```
Please move to a separate pr. This is not related to the testnet.

itamararjuan replied:
Moved already and also merged to master. Will be deleted from here.

itamararjuan commented:
I've addressed the concerns that @netoneko had with this PR.
When you are ready to review let me know so we can proceed with merging this PR to master

Preferrably before Thursday if possible :-)

netoneko commented on .circleci/check-testnet-deployment.js:
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env node
+
+const { waitUntilSync, getBlockHeight, getCommit } = require('@orbs-network/orbs-nebula/lib/metrics');
```
We decided to unpublish `@orbs-network/orbs-nebula` a month ago. Why is it used here?

itamararjuan replied:
In your previous review you wanted me to use your code from Nebula
This is the official package. We talked about unpublish (and we did) the other package (`orbs-nebula`)

This scoped package is the official one which we left

netoneko replied:
Official docs still use `npm install -g orbs-nebula`

Also maybe it would be easier to just install from our github master? We rarely release new nebula version.

netoneko commented on .circleci/check-testnet-deployment.js:
```diff
@@ -0,0 +1,99 @@
+    return topology;
+}
+
+async function eventuallyDeployed({ chainId, nodes }) {
```
`waitUnitlCommit` from `metrics` does that already: https://github.com/orbs-network/nebula/blob/master/lib/metrics.js#L112

itamararjuan replied:
I do not agree. I want to see all nodes have updated to the new version. This is not the same.
Your method is single endpoint centric

netoneko replied:
You can run it multiple times waiting for the same commit. Same effect, less code.

netoneko commented on .circleci/check-testnet-deployment.js:
```diff
@@ -0,0 +1,99 @@
+    const currentBlockheight = await getBlockHeight(firstEndpoint);
+
+    try {
+        await waitUntilSync(firstEndpoint, currentBlockheight + 50);
```
50 is too long, blocks are closed at a rate of 9 seconds on an idle system. It would be 7 minutes after we waited for the network to start producing blocks, which itself can take 10-15 minutes.

itamararjuan replied:
I'm updating it to `10`

netoneko commented on .circleci/check-testnet-deployment.js:
```diff
@@ -0,0 +1,99 @@
+    if (results.filter(r => r.ok === true).length === chains.length) {
+        console.log('New version deployed successfully on all chains in the testnet');
+    } else {
+        console.error('New version was not deployed on all nodes within the defined 5 minutes window, quiting..');
```
5 minutes is too low. It will never work that fast.

itamararjuan replied:
I've updated it to 15 minutes

netoneko replied:
Cool.

netoneko commented on .circleci/check-testnet-deployment.js:
```diff
@@ -0,0 +1,99 @@
+}
+
+async function getTopology() {
+    const result = await fetch(TOPOLOGY);
```
You have already downloaded it with curl, why do it here again? I would prefer modifying the local file.

itamararjuan replied:
Agreed - changing it.

netoneko commented on .circleci/package.json:
```diff
@@ -0,0 +1,13 @@
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "@orbs-network/orbs-nebula": "^1.1.2",
```
Should be just `orbs-nebula`.

netoneko commented on .circleci/check-testnet-deployment.js:
```diff
@@ -0,0 +1,91 @@
+    };
+}
+
+async function eventuallyClosingBlocks({ chainId, nodes }) {
```
Same can be done with nebula call for status, waiting for all of them to become green.
